src/main/java/me/itzg/tsdbcassandra/services/DownsampleProcessor.java

```diff
@@ -0,0 +1,216 @@
+    )
+        .publishOn(scheduler);
+
+    final Flux<Tuple2<DataDownsampled, Boolean>> aggregated = aggregateData(data,
+        pendingDownsampleSet.getTenant(),
+        pendingDownsampleSet.getSeriesSet(),
+        downsampleProperties.getGranularities().iterator(), isCounter
+    );
+
+    return
+        aggregated
+            .then(
+                metadataService.updateMetricNames(
+                    pendingDownsampleSet.getTenant(),
+                    seriesSetService.metricNameFromSeriesSet(pendingDownsampleSet.getSeriesSet()),
+                    isCounter ? Set.of(Aggregator.sum) : Set.of(Aggregator.sum, Aggregator.min, Aggregator.max, Aggregator.avg)
+                )
+            )
+            .then(
+                downsampleTrackingService.complete(pendingDownsampleSet)
+            )
+            .doOnError(throwable -> log.warn("Failed to downsample {}", pendingDownsampleSet, throwable))
+            .doOnSuccess(o -> log.debug("Completed downsampling of {}", pendingDownsampleSet));
+  }
+
+  /**
+   * Aggregates the given data into the next granularity, schedules the storage of that data,
+   * and recurses until the remaining granularities are processed.
+   * @param data a flux of either raw {@link me.itzg.tsdbcassandra.downsample.SingleValueSet}s or
+   * aggregated {@link AggregatedValueSet}s from the prior granularity.
+   * @param tenant the tenant of the pending downsample set
+   * @param seriesSet the series-set of the pending downsample set
+   * @param granularities remaining granularties to process
+   * @param isCounter indicates if the original metric is a counter or gauge
+   * @return a flux of the stored downsamples along with the "applied" indicator from Cassandra
+   */
+  public Flux<Tuple2<DataDownsampled, Boolean>> aggregateData(Flux<? extends ValueSet> data,
+                                                              String tenant,
+                                                              String seriesSet,
+                                                              Iterator<Granularity> granularities,
+                                                              boolean isCounter) {
+    if (!granularities.hasNext()) {
+      // end of the recursion so pop back out
+      return Flux.empty();
+    }
+
+    final Granularity granularity = granularities.next();
+    final TemporalNormalizer normalizer = new TemporalNormalizer(granularity.getWidth());
+
+    final Flux<AggregatedValueSet> aggregated =
+        data
+            .doOnNext(valueSet -> log.trace("Aggregating {} into granularity={}", valueSet, granularity))
+            // group the incoming data by granularity-time-window
+            .windowUntilChanged(
+                valueSet -> valueSet.getTimestamp().with(normalizer), Instant::equals)
+            // ...and then do the aggregation math on those
+            .concatMap(valueSetFlux -> valueSetFlux.collect(
```

Comment:
A premature optimization, but it seems collectMap() will impose a lot of probably unnecessary creation of publishers just to do basic arithmetic.   I wonder if it will be better to transform this to an iterable first, do the math and create more fluxes from the result.

Nothing I think we should do now.  Just thinking that this is the computationally intense part of rollups and we probably will want to optimize this area.

Reply:
Definitely a valid spot to confirm efficiency. I am assuming the repeated calls into the aggregating collector are equivalent to an iterable-stream; however, I'll poke around in the call stack to better understand what it actually does.

Reply:
...from breakpoint stack traces, it seems to use a `MonoStreamCollector` for each windowed flux and spends most time calling my `ValueSetCollector` to accumulate:

https://github.com/reactor/reactor-core/blob/3.3.x/reactor-core/src/main/java/reactor/core/publisher/MonoStreamCollector.java#L125

and then the finisher at

https://github.com/reactor/reactor-core/blob/3.3.x/reactor-core/src/main/java/reactor/core/publisher/MonoStreamCollector.java#L159

Reply:
...exploring the mysteries of `Publisher`s a bit more, it really just boils down to an interface exposing a single `subscribe` method

https://www.reactive-streams.org/reactive-streams-1.0.3-javadoc/org/reactivestreams/Publisher.html

which accepts a `Subscriber` and it in turn is a simple interface of something that can accept signals like onNext, onComplete, etc:

https://www.reactive-streams.org/reactive-streams-1.0.3-javadoc/org/reactivestreams/Subscriber.html

Reply:
cool, ok that doesn't sound too bad